components/codec/src/columnar/bool.rs

```diff
@@ -0,0 +1,245 @@
+        } else {
+            Self::encode_without_compression(buf, values)
+        }
+    }
+
+    fn estimated_encoded_size<I>(&self, values: I) -> usize
+    where
+        I: Iterator<Item = bool>,
+    {
+        let num_values = values.count();
+        if Self::need_compress(num_values) {
+            BitSet::num_bytes(num_values)
+                + Self::COMPRESSION_SIZE
+                + Self::NUM_VALUES_SIZE
+                + Self::VERSION_SIZE
+        } else {
+            num_values + Self::VERSION_SIZE + Self::COMPRESSION_SIZE
+        }
+    }
+
+    fn decode<B, F>(&self, buf: &mut B, f: F) -> Result<()>
+    where
+        B: Buf,
+        F: FnMut(bool) -> Result<()>,
+    {
+        let buf = buf.chunk();
+        ensure!(
+            buf.len() > Self::VERSION_SIZE + Self::COMPRESSION_SIZE,
+            NotEnoughBytes { len: buf.len() }
+        );
+
+        // Decode the version.
+        let version = buf[0];
+        ensure!(version == Self::VERSION, InvalidVersion { version });
+
+        // Decode the compression.
+        let compression_index = buf.len() - 1;
+        match Self::decode_compression(buf[compression_index])? {
+            Compression::None => Self::decode_without_compression(buf, f)?,
+            Compression::BitSet => Self::decode_with_compression(buf, f)?,
+        }
+
+        Ok(())
+    }
+
+    fn encode_without_compression<B, I>(buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = bool>,
+    {
+        for v in values {
+            buf.put_u8(v as u8);
+        }
+
+        buf.put_u8(Compression::None as u8);
+
+        Ok(())
+    }
+
+    fn decode_without_compression<F>(buf: &[u8], mut f: F) -> Result<()>
+    where
+        F: FnMut(bool) -> Result<()>,
+    {
+        let data_block_start = Self::VERSION_SIZE;
+        let data_block_end = buf.len() - Self::COMPRESSION_SIZE;
+        let data_block = &buf[data_block_start..data_block_end];
+        for v in data_block {
+            match *v {
+                0 => f(false),
+                1 => f(true),
+                _ => InvalidBooleanValue { value: *v }.fail(),
+            }?
+        }
+
+        Ok(())
+    }
+
+    fn encode_with_compression<B, I>(buf: &mut B, num_values: usize, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = bool>,
+    {
+        buf.put_u32(num_values as u32);
+
+        let mut one_byte_bits = [false; 8];
+        let mut offset = 0;
+        for v in values {
+            one_byte_bits[offset] = v;
+            offset += 1;
+            if offset == 8 {
+                let bit_set = BitSet::one_byte(&one_byte_bits);
+                buf.put_u8(bit_set);
+
+                // Reset the offset and the bits buf.
+                offset = 0;
+                one_byte_bits = [false; 8];
+            }
+        }
+
+        // Put the remaining bits.
+        if offset > 0 {
```

Comment:
what if last bit is 0, will this lost one element?

Reply:
This is offset, the last element has been encoded into the buf if it is equal 0.